File: lib/scrapers/media_helpers.dart
import '../models/filter_settings.dart';
import '../models/media_item.dart';

const videoExtensions = {'mp4', 'webm'};
const imageExtensions = {'jpg', 'jpeg', 'png', 'gif', 'webp'};
const supportedMediaExtensions = {...videoExtensions, ...imageExtensions};

String? mediaExtensionFromUrl(String url) {
  final uri = Uri.tryParse(url);
  final path = uri?.path.toLowerCase() ?? url.toLowerCase();
  final match = RegExp(r'\.([a-z0-9]+)$').firstMatch(path);
  final extension = match?.group(1);
  if (extension == null) return null;
  return supportedMediaExtensions.contains(extension) ? extension : null;
}

String fileNameFromUrl(String url, {String fallback = 'media'}) {
  final uri = Uri.tryParse(url);
  final segment = uri?.pathSegments.where((part) => part.isNotEmpty).lastOrNull;
  if (segment == null || segment.trim().isEmpty) return fallback;
  return Uri.decodeComponent(segment);
}

String? qualityFromFileName(String fileName) {
  final match = RegExp(
    r'(?<!\d)(\d{3,4}p)(?!\d)',
    caseSensitive: false,
  ).firstMatch(fileName);
  return match?.group(1)?.toLowerCase();
}

List<MediaItem> applyMediaFilters(
  List<MediaItem> items,
  FilterSettings filters,
) {
  return items
      .where((item) {
        return filters.acceptsExtension(item.extension) &&
            filters.acceptsSize(item.sizeBytes);
      })
      .toList(growable: false);
}

extension FirstOrNullExtension<T> on Iterable<T> {
  T? get lastOrNull {
    final iterator = this.iterator;
    T? last;
    while (iterator.moveNext()) {
      last = iterator.current;
    }
    return last;
  }
}

File: app/src/main/java/io/github/rustnomicon/sourceinstaller/scrape/MediaHelpers.kt
package io.github.rustnomicon.sourceinstaller.scrape

import io.github.rustnomicon.sourceinstaller.data.model.FilterSettings
import io.github.rustnomicon.sourceinstaller.data.model.MediaItem
import java.net.URI
import java.net.URLDecoder
import java.nio.charset.StandardCharsets

val VIDEO_EXTENSIONS: Set<String> = setOf("mp4", "webm")
val IMAGE_EXTENSIONS: Set<String> = setOf("jpg", "jpeg", "png", "gif", "webp")
val SUPPORTED_MEDIA_EXTENSIONS: Set<String> = VIDEO_EXTENSIONS + IMAGE_EXTENSIONS

private val EXTENSION_AT_END = Regex("""\.([a-z0-9]+)$""")
private val QUALITY_MARKER = Regex("""(?<!\d)(\d{3,4}p)(?!\d)""", RegexOption.IGNORE_CASE)

fun mediaExtensionFromUrl(url: String): String? {
    val path = try {
        URI(url).path?.lowercase() ?: url.lowercase()
    } catch (_: Exception) {
        url.lowercase()
    }
    val extension = EXTENSION_AT_END.find(path)?.groupValues?.get(1) ?: return null
    return if (extension in SUPPORTED_MEDIA_EXTENSIONS) extension else null
}

fun fileNameFromUrl(url: String, fallback: String = "media"): String {
    val segment = try {
        URI(url).rawPath
            ?.split('/')
            ?.lastOrNull { it.isNotEmpty() }
    } catch (_: Exception) {
        null
    } ?: return fallback

    return try {
        URLDecoder.decode(segment, StandardCharsets.UTF_8)
    } catch (_: Exception) {
        segment
    }
}

fun qualityFromFileName(fileName: String): String? =
    QUALITY_MARKER.find(fileName)?.groupValues?.get(1)?.lowercase()

fun applyMediaFilters(
    items: List<MediaItem>,
    filters: FilterSettings,
): List<MediaItem> = items.filter {
    filters.acceptsExtension(it.extension) && filters.acceptsSize(it.sizeBytes)
}
